bliss/models/location_encoder.py

```diff
@@ -479,3 +702,91 @@ def forward(self, x):
 
         out = x + identity
         return F.relu(out, inplace=True)
+
+
+def _get_log_probs_all_perms(
+    locs_log_probs_all,
+    star_params_log_probs_all,
+    true_galaxy_bools,
+    is_on_array,
+):
+    # get log-probability under every possible matching of estimated source to true source
+    n_ptiles = star_params_log_probs_all.size(0)
+    max_detections = star_params_log_probs_all.size(-1)
+
+    n_permutations = math.factorial(max_detections)
+    locs_log_probs_all_perm = torch.zeros(
+        n_ptiles, n_permutations, device=locs_log_probs_all.device
+    )
+    star_params_log_probs_all_perm = locs_log_probs_all_perm.clone()
+
+    for i, perm in enumerate(itertools.permutations(range(max_detections))):
+        # note that we multiply is_on_array, we only evaluate the loss if the source is on.
+        locs_log_probs_all_perm[:, i] = (
+            locs_log_probs_all[:, perm].diagonal(dim1=1, dim2=2) * is_on_array
+        ).sum(1)
+
+        # if star, evaluate the star parameters,
+        # hence the multiplication by (1 - true_galaxy_bools)
+        # the diagonal is a clever way of selecting the elements of each permutation (first index
+        # of mean/var with second index of true_param etc.)
+        star_params_log_probs_all_perm[:, i] = (
+            star_params_log_probs_all[:, perm].diagonal(dim1=1, dim2=2)
+            * is_on_array
+            * (1 - true_galaxy_bools)
+        ).sum(1)
+
+    return locs_log_probs_all_perm, star_params_log_probs_all_perm
+
+
+def get_min_perm_loss(
```

Comment:
I noticed that you made some methods private, I think you can do the same with this one and the next one

Reply:
Done